src/transformers/integrations/bitsandbytes.py

```diff
@@ -363,13 +363,14 @@ def dequantize_bnb_weight(weight: "torch.nn.Parameter", dtype: "torch.dtype", st
     if state.SCB is None:
         state.SCB = weight.SCB
 
-    im = torch.eye(weight.data.shape[-1]).contiguous().half().to(weight.device)
-    im, imt, SCim, SCimt, coo_tensorim = bnb.functional.double_quant(im)
-    im, Sim = bnb.functional.transform(im, "col32")
-    if state.CxB is None:
-        state.CxB, state.SB = bnb.functional.transform(weight.data, to_order=state.formatB)
-    out32, Sout32 = bnb.functional.igemmlt(im, state.CxB, Sim, state.SB)
-    return bnb.functional.mm_dequant(out32, Sout32, SCim, state.SCB, bias=None).t().to(dtype)
+    if hasattr(bnb.functional, "int8_vectorwise_dequant"):
+        # Use bitsandbytes API if available (requires v0.45.0+)
+        dequantized = bnb.functional.int8_vectorwise_dequant(weight.data, state.SCB)
+    else:
+        # Multiply by (scale/127) to dequantize.
+        dequantized = weight.data * state.SCB.view(-1, 1) * 7.874015718698502e-3
```

Comment:
This line equivalent to what we had before ?

Reply:
Yes. It is also the same way the selected 8bit weights are dequantized for fp16 compute in the LLM.int8() implementation: https://github.com/bitsandbytes-foundation/bitsandbytes/blob/7dca70049566b5b1c55cbd67e1cb191729a98152/bitsandbytes/autograd/_functions.py#L380

The method that this replaces was multiplying the quantized weights by a quantized identity matrix (int8 * int8 -> int32):
$W_{INT32} = I_{INT8}*W_{INT8} $

The dequantization feature `mm_dequant` was intended for dequantizing int32 matmul results, and as such performs $W_{FP16} = (W_{INT32} * (SC_{im} \bigotimes SCB)) / (127 * 127)$.

![image](https://github.com/user-attachments/assets/ac6dfd4c-8f9e-473d-a9cb-b2b714d07351)

In this new version, I'm avoiding the cost of creating the identity matrix, quantizing it, and performing the matmul realizing int32 values. Mathematically this is equivalent, and more importantly it's simpler and introduces less error.

Reply:
Nice !